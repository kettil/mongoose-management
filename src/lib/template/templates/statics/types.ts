/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/naming-convention */
/*
 * ######################################################################
 * #                                                                    #
 * #                       Do not change the file!                      #
 * #                                                                    #
 * # All changes are deleted the next time the collection data is read. #
 * #                                                                    #
 * ######################################################################
 */

import { IndexOptions } from 'mongodb';
import { Document, DocumentQuery, Aggregate, Model } from 'mongoose';

// ##### Virtuals #####

/**
 * Helpers
 */
type Keys<T> = { [K in keyof T]: T[K] extends never ? never : K }[keyof T];
type Diff<T, U> = T extends U ? never : T;

/**
 * Extracts the type from the function
 */
type VirtualFunctionTypeGet<T extends { [k: string]: any }> = T['get'] extends () => void
  ? ReturnType<T['get']>
  : never;
type VirtualFunctionTypeSet<T extends { [k: string]: any }> = T['set'] extends (value: infer A) => void ? A : never;

/**
 * List of all types (with the type "never")
 */
type VirtualFunctionTypeGetList<T> = { [K in keyof T]: VirtualFunctionTypeGet<T[K]> };
type VirtualFunctionTypeSetList<T> = { [K in keyof T]: VirtualFunctionTypeSet<T[K]> };

/**
 * List of all types (without the type "never")
 */
type VirtualTypeGetList<T> = Pick<VirtualFunctionTypeGetList<T>, Keys<VirtualFunctionTypeGetList<T>>>;
type VirtualTypeSetList<T> = Pick<VirtualFunctionTypeSetList<T>, Keys<VirtualFunctionTypeSetList<T>>>;

/**
 * List of Keys where the object have only "get"
 */
type VirtualKeysGetOnly<T> = Diff<keyof VirtualTypeGetList<T>, keyof VirtualTypeSetList<T>>;

/**
 * List of Keys where without "get only" objects
 */
type VirtualKeysBoth<T> = Diff<keyof T, VirtualKeysGetOnly<T>>;

type VirtualSet<T> = { [K in keyof VirtualTypeSetList<T>]: VirtualTypeSetList<T>[K] };
type VirtualGet<T> = { [K in keyof VirtualTypeGetList<T> & VirtualKeysBoth<T>]: VirtualTypeGetList<T>[K] };
type VirtualOnlyGet<T> = {
  readonly [K in keyof VirtualTypeGetList<T> & VirtualKeysGetOnly<T>]: VirtualTypeGetList<T>[K];
};

export type VirtualType<T> = VirtualOnlyGet<T> & VirtualGet<T> & VirtualSet<T>;

// ##### Hooks #####

type HooksMethodDocument = 'validate' | 'save' | 'remove';
type HooksMethodQueryWithout = 'update' | 'updateOne' | 'updateMany' | 'deleteOne' | 'deleteMany';
type HooksMethodQueryCount = 'countDocuments';
type HooksMethodQuerySingle = 'findOne' | 'findOneAndDelete' | 'findOneAndRemove' | 'findOneAndUpdate';
type HooksMethodQueryMultiple = 'find';
type HooksMethodAggreagate = 'aggregate';
type HooksMethodInsert = 'insertMany';

type HooksReturn = Promise<void> | void;

type HooksPre<
  D extends Document,
  Q extends DocumentQuery<D | D[], D>,
  A extends Aggregate<D>,
  M extends Model<D>,
  E extends Record<string, unknown>
> = {
  (method: 'init', cb: (this: D) => void): void;
  (method: HooksMethodDocument, cb: (this: D) => HooksReturn): void;
  (method: HooksMethodQueryCount, cb: (this: Q & E) => HooksReturn): void;
  (method: HooksMethodQueryMultiple, cb: (this: Q & E) => HooksReturn): void;
  (method: HooksMethodQuerySingle, cb: (this: Q & E) => HooksReturn): void;
  (method: HooksMethodQueryWithout, cb: (this: Q & E) => HooksReturn): void;
  (method: HooksMethodAggreagate, cb: (this: A) => HooksReturn): void;
  (method: HooksMethodInsert, cb: (this: M) => HooksReturn): void;
};

type HooksPost<
  D extends Document,
  Q extends DocumentQuery<D | D[], D>,
  A extends Aggregate<D>,
  M extends Model<D>,
  E extends Record<string, unknown>
> = {
  (method: 'init', cb: (this: D, doc: D) => void): void;
  (method: HooksMethodDocument, cb: (this: D, doc: D) => HooksReturn): void;
  (method: HooksMethodQueryCount, cb: (this: Q & E, count: number) => HooksReturn): void;
  (method: HooksMethodQueryMultiple, cb: (this: Q & E, docs: D[]) => HooksReturn): void;
  (method: HooksMethodQuerySingle, cb: (this: Q & E, doc: D) => HooksReturn): void;
  (method: HooksMethodQueryWithout, cb: (this: Q & E) => HooksReturn): void;
  (method: HooksMethodAggreagate, cb: (this: A, docs: D[]) => HooksReturn): void;
  (method: HooksMethodInsert, cb: (this: M, docs: D[]) => HooksReturn): void;
};

type ClearExtendQuery<
  D extends Document,
  Q extends DocumentQuery<D | D[], D>,
  E extends Record<string, unknown>
> = Partial<Omit<E, keyof Q>>;

export type HooksHandler<
  D extends Document,
  Q extends DocumentQuery<D | D[], D>,
  A extends Aggregate<D>,
  M extends Model<D>,
  ExtendQuery extends Record<string, unknown>
> = (
  pre: HooksPre<D, Q, A, M, ClearExtendQuery<D, Q, ExtendQuery>>,
  post: HooksPost<D, Q, A, M, ClearExtendQuery<D, Q, ExtendQuery>>,
) => void;

// ##### Others #####

export type ExtendTimestampType = {
  readonly createdAt: Date;
  readonly updatedAt: Date;
};

export type ExtendIdType = {
  id: any;
};

export type ExportType<T extends Record<string, unknown>> = T &
  ExtendTimestampType &
  ExtendIdType & {
    _id: unknown;
    __v: unknown;
  };

type IndexOptionsExclude = 'dropDups' | 'min' | 'max' | 'v' | 'session' | 'w' | 'j' | 'wtimeout';

export type IndexType = {
  fields: { [k: string]: 1 | -1 | 'text' | '2dsphere' | 'hashed' };
  options: Pick<IndexOptions, Exclude<keyof IndexOptions, IndexOptionsExclude>>;
};
