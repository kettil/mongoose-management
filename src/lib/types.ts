import { Separator } from 'inquirer';

import Prompts from './prompts';
import Storage from './storage';
import Create from './template/create';

import * as mongoTypes from './mongo';

/**
 *
 */
export type objType<K extends string | number | symbol = string, T = any> = Record<K, T>;

/**
 *
 */
export type levelOptionsType = {
  prompts: Prompts;
  storage: Storage;
  creater: Create;
};

/**
 *
 */
export type schemaType = keyof typeof mongoTypes.schemaTypes;

/**
 *
 */
export type schemaNormalType = keyof typeof mongoTypes.schemaTypesNormal;

/**
 *
 */
export type schemaIndexType = keyof typeof mongoTypes.schemaIndexTypes;

/**
 *
 */
export type dataType = {
  groups: dataGroupType[];
};

/**
 *
 */
export type dataGroupType = {
  path: string;
  collections: dataCollectionType[];
};

/**
 *
 */
export type dataCollectionType = {
  name: string;
  idType: schemaType;
  columns: dataColumnType[];
  indexes: dataIndexType[];
};

/**
 *
 */
export type dataColumnType = {
  name: string;
  type: schemaType;

  required?: boolean;
  default?: string | number;

  // string
  lowercase?: boolean;
  uppercase?: boolean;
  trim?: boolean;
  match?: string;
  enum?: string;
  minLength?: number;
  maxLength?: number;

  // number
  min?: number;
  max?: number;
} & dataColumnInternalValuesType;

/**
 *
 */
export type dataColumnInternalValuesType = {
  // objectId
  populate?: string;

  subTypes?: schemaNormalType[];
  subColumns?: dataColumnType[];
};

/**
 *
 */
export type dataIndexType = {
  name: string;
  columns: { [k: string]: dataIndexColumnValueType };
  properties: {
    unique?: boolean;
    sparse?: boolean;
  };
} & dataIndexInternalValuesType;

/**
 *
 */
export type dataIndexInternalValuesType = {
  readonly?: boolean;
};

/**
 *
 */
export type dataIndexColumnValueType = 1 | -1 | 'text' | 'hashed' | '2dsphere';

/**
 *
 */
export type choicesType<T> = choiceType<T> | InstanceType<typeof Separator>;

/**
 *
 */
export type choiceType<T> = choiceListType<choiceValueType<T>>;

/**
 *
 */
export type choiceListType<T> = {
  name: string;
  value: T;
  short: string;
};

/**
 *
 */
export type choiceValueType<T> = {
  action?:
    | 'create'
    | 'createColumn'
    | 'createIndex'
    | 'back'
    | 'backToCollection'
    | 'exit'
    | 'save'
    | 'edit'
    | 'remove'
    | 'write';
  data?: T;
};

/**
 *
 */
export type templateTypesType = 'index' | 'documents' | 'interfaces' | 'models' | 'repositories';

/**
 *
 */
export type templateCollectionNamesType = {
  collectionNameLower: string;
  collectionNameUpper: string;
  collectionNameRaw: string;
};

/**
 *
 */
export type templateCollectionType = {
  interfaceName: string;

  schemaDefinitions: string;
  SchemaIndexes: string;
  schemaTypes: string;

  additionalImports: string[];
} & templateCollectionNamesType;

export type Unpacked<T> = T extends Array<infer U> ? U : T;

export type OptionalValues<T> = T extends { [K in keyof T]: any } ? { [K in keyof T]?: T[K] } : T;
