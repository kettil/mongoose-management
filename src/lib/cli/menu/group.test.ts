jest.mock('../../prompts');

import Prompts from '../../prompts';
import GroupDataset from '../dataset/group';

import GroupMenu from './group';

import { choicesType } from '../../types';

describe('Check the GroupMenu class', () => {
  let group: GroupDataset;
  let prompts: any;
  let menu: any;

  beforeEach(() => {
    group = new GroupDataset(
      {
        path: 'path',
        collections: [
          { name: 'collection2', columns: [], indexes: [] },
          { name: 'collection1', columns: [], indexes: [] },
        ],
      },
      jest.fn() as any,
    );
    group.setReference();

    prompts = new Prompts();

    menu = new GroupMenu(prompts);
  });

  test('initialize the class', () => {
    expect(menu).toBeInstanceOf(GroupMenu);

    expect(menu.prompts).toBe(prompts);
  });

  test('it should be return menu selection when exec() is called with collections', async () => {
    ((prompts.menu as any) as jest.Mock).mockImplementation((_: string, choices: choicesType<any>[]) => {
      expect(choices).toMatchSnapshot();

      return { action: 'edit', data: undefined };
    });

    const result = await menu.exec(group);

    expect(result).toEqual({ action: 'edit', data: undefined });
    expect(prompts.menu).toHaveBeenCalledTimes(1);
    expect(prompts.menu).toHaveBeenCalledWith('Choose a collection or a command:', expect.any(Array));
  });

  test('it should be return menu selection when exec() is called without collections', async () => {
    ((prompts.menu as any) as jest.Mock).mockImplementation((_: string, choices: choicesType<any>[]) => {
      expect(choices).toMatchSnapshot();

      return { action: 'edit', data: undefined };
    });

    group.getCollections().forEach((c) => c.remove());

    const result = await menu.exec(group);

    expect(result).toEqual({ action: 'edit', data: undefined });
    expect(prompts.menu).toHaveBeenCalledTimes(1);
    expect(prompts.menu).toHaveBeenCalledWith('Choose a collection or a command:', expect.any(Array));
  });

  test('it should be return adjusted collections when getChoiceList() is called', () => {
    const result = menu.getChoiceList(group.getCollections());

    expect(result).toMatchSnapshot();
  });
});
