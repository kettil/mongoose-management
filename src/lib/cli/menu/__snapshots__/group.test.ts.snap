// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check the GroupMenu class it should be return adjusted collections when getChoiceList() is called 1`] = `
Array [
  Object {
    "name": "collection2",
    "short": "collection2",
    "value": Object {
      "data": CollectionDataset {
        "columns": Array [
          ColumnDataset {
            "collection": [Circular],
            "column": Object {
              "name": "_id",
              "type": "objectId",
            },
            "columns": Array [],
            "index": undefined,
            "parent": [Circular],
            "readonly": true,
            "subTypes": Array [],
          },
          ColumnDataset {
            "collection": [Circular],
            "column": Object {
              "name": "createdAt",
              "type": "date",
            },
            "columns": Array [],
            "index": undefined,
            "parent": [Circular],
            "readonly": true,
            "subTypes": Array [],
          },
          ColumnDataset {
            "collection": [Circular],
            "column": Object {
              "name": "updatedAt",
              "type": "date",
            },
            "columns": Array [],
            "index": undefined,
            "parent": [Circular],
            "readonly": true,
            "subTypes": Array [],
          },
        ],
        "idType": "objectId",
        "indexes": Array [],
        "name": "collection2",
        "parent": GroupDataset {
          "collections": Array [
            [Circular],
            CollectionDataset {
              "columns": Array [
                ColumnDataset {
                  "collection": [Circular],
                  "column": Object {
                    "name": "_id",
                    "type": "objectId",
                  },
                  "columns": Array [],
                  "index": undefined,
                  "parent": [Circular],
                  "readonly": true,
                  "subTypes": Array [],
                },
                ColumnDataset {
                  "collection": [Circular],
                  "column": Object {
                    "name": "createdAt",
                    "type": "date",
                  },
                  "columns": Array [],
                  "index": undefined,
                  "parent": [Circular],
                  "readonly": true,
                  "subTypes": Array [],
                },
                ColumnDataset {
                  "collection": [Circular],
                  "column": Object {
                    "name": "updatedAt",
                    "type": "date",
                  },
                  "columns": Array [],
                  "index": undefined,
                  "parent": [Circular],
                  "readonly": true,
                  "subTypes": Array [],
                },
              ],
              "idType": "objectId",
              "indexes": Array [],
              "name": "collection1",
              "parent": [Circular],
              "specialColumns": Array [
                Array [
                  "createdAt",
                  "date",
                ],
                Array [
                  "updatedAt",
                  "date",
                ],
                Array [
                  "_id",
                  "objectId",
                ],
              ],
            },
          ],
          "parent": [MockFunction],
          "path": "path",
        },
        "specialColumns": Array [
          Array [
            "createdAt",
            "date",
          ],
          Array [
            "updatedAt",
            "date",
          ],
          Array [
            "_id",
            "objectId",
          ],
        ],
      },
    },
  },
  Object {
    "name": "collection1",
    "short": "collection1",
    "value": Object {
      "data": CollectionDataset {
        "columns": Array [
          ColumnDataset {
            "collection": [Circular],
            "column": Object {
              "name": "_id",
              "type": "objectId",
            },
            "columns": Array [],
            "index": undefined,
            "parent": [Circular],
            "readonly": true,
            "subTypes": Array [],
          },
          ColumnDataset {
            "collection": [Circular],
            "column": Object {
              "name": "createdAt",
              "type": "date",
            },
            "columns": Array [],
            "index": undefined,
            "parent": [Circular],
            "readonly": true,
            "subTypes": Array [],
          },
          ColumnDataset {
            "collection": [Circular],
            "column": Object {
              "name": "updatedAt",
              "type": "date",
            },
            "columns": Array [],
            "index": undefined,
            "parent": [Circular],
            "readonly": true,
            "subTypes": Array [],
          },
        ],
        "idType": "objectId",
        "indexes": Array [],
        "name": "collection1",
        "parent": GroupDataset {
          "collections": Array [
            CollectionDataset {
              "columns": Array [
                ColumnDataset {
                  "collection": [Circular],
                  "column": Object {
                    "name": "_id",
                    "type": "objectId",
                  },
                  "columns": Array [],
                  "index": undefined,
                  "parent": [Circular],
                  "readonly": true,
                  "subTypes": Array [],
                },
                ColumnDataset {
                  "collection": [Circular],
                  "column": Object {
                    "name": "createdAt",
                    "type": "date",
                  },
                  "columns": Array [],
                  "index": undefined,
                  "parent": [Circular],
                  "readonly": true,
                  "subTypes": Array [],
                },
                ColumnDataset {
                  "collection": [Circular],
                  "column": Object {
                    "name": "updatedAt",
                    "type": "date",
                  },
                  "columns": Array [],
                  "index": undefined,
                  "parent": [Circular],
                  "readonly": true,
                  "subTypes": Array [],
                },
              ],
              "idType": "objectId",
              "indexes": Array [],
              "name": "collection2",
              "parent": [Circular],
              "specialColumns": Array [
                Array [
                  "createdAt",
                  "date",
                ],
                Array [
                  "updatedAt",
                  "date",
                ],
                Array [
                  "_id",
                  "objectId",
                ],
              ],
            },
            [Circular],
          ],
          "parent": [MockFunction],
          "path": "path",
        },
        "specialColumns": Array [
          Array [
            "createdAt",
            "date",
          ],
          Array [
            "updatedAt",
            "date",
          ],
          Array [
            "_id",
            "objectId",
          ],
        ],
      },
    },
  },
]
`;

exports[`Check the GroupMenu class it should be return menu selection when exec() is called with collections 1`] = `
Array [
  Separator {
    "line": "[2mGroup: [0m[1mpath[2m[0m[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
  Separator {
    "line": "[2mCollections list[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
  Object {
    "name": "collection2",
    "short": "collection2",
    "value": Object {
      "data": CollectionDataset {
        "columns": Array [
          ColumnDataset {
            "collection": [Circular],
            "column": Object {
              "name": "_id",
              "type": "objectId",
            },
            "columns": Array [],
            "index": undefined,
            "parent": [Circular],
            "readonly": true,
            "subTypes": Array [],
          },
          ColumnDataset {
            "collection": [Circular],
            "column": Object {
              "name": "createdAt",
              "type": "date",
            },
            "columns": Array [],
            "index": undefined,
            "parent": [Circular],
            "readonly": true,
            "subTypes": Array [],
          },
          ColumnDataset {
            "collection": [Circular],
            "column": Object {
              "name": "updatedAt",
              "type": "date",
            },
            "columns": Array [],
            "index": undefined,
            "parent": [Circular],
            "readonly": true,
            "subTypes": Array [],
          },
        ],
        "idType": "objectId",
        "indexes": Array [],
        "name": "collection2",
        "parent": GroupDataset {
          "collections": Array [
            [Circular],
            CollectionDataset {
              "columns": Array [
                ColumnDataset {
                  "collection": [Circular],
                  "column": Object {
                    "name": "_id",
                    "type": "objectId",
                  },
                  "columns": Array [],
                  "index": undefined,
                  "parent": [Circular],
                  "readonly": true,
                  "subTypes": Array [],
                },
                ColumnDataset {
                  "collection": [Circular],
                  "column": Object {
                    "name": "createdAt",
                    "type": "date",
                  },
                  "columns": Array [],
                  "index": undefined,
                  "parent": [Circular],
                  "readonly": true,
                  "subTypes": Array [],
                },
                ColumnDataset {
                  "collection": [Circular],
                  "column": Object {
                    "name": "updatedAt",
                    "type": "date",
                  },
                  "columns": Array [],
                  "index": undefined,
                  "parent": [Circular],
                  "readonly": true,
                  "subTypes": Array [],
                },
              ],
              "idType": "objectId",
              "indexes": Array [],
              "name": "collection1",
              "parent": [Circular],
              "specialColumns": Array [
                Array [
                  "createdAt",
                  "date",
                ],
                Array [
                  "updatedAt",
                  "date",
                ],
                Array [
                  "_id",
                  "objectId",
                ],
              ],
            },
          ],
          "parent": [MockFunction],
          "path": "path",
        },
        "specialColumns": Array [
          Array [
            "createdAt",
            "date",
          ],
          Array [
            "updatedAt",
            "date",
          ],
          Array [
            "_id",
            "objectId",
          ],
        ],
      },
    },
  },
  Object {
    "name": "collection1",
    "short": "collection1",
    "value": Object {
      "data": CollectionDataset {
        "columns": Array [
          ColumnDataset {
            "collection": [Circular],
            "column": Object {
              "name": "_id",
              "type": "objectId",
            },
            "columns": Array [],
            "index": undefined,
            "parent": [Circular],
            "readonly": true,
            "subTypes": Array [],
          },
          ColumnDataset {
            "collection": [Circular],
            "column": Object {
              "name": "createdAt",
              "type": "date",
            },
            "columns": Array [],
            "index": undefined,
            "parent": [Circular],
            "readonly": true,
            "subTypes": Array [],
          },
          ColumnDataset {
            "collection": [Circular],
            "column": Object {
              "name": "updatedAt",
              "type": "date",
            },
            "columns": Array [],
            "index": undefined,
            "parent": [Circular],
            "readonly": true,
            "subTypes": Array [],
          },
        ],
        "idType": "objectId",
        "indexes": Array [],
        "name": "collection1",
        "parent": GroupDataset {
          "collections": Array [
            CollectionDataset {
              "columns": Array [
                ColumnDataset {
                  "collection": [Circular],
                  "column": Object {
                    "name": "_id",
                    "type": "objectId",
                  },
                  "columns": Array [],
                  "index": undefined,
                  "parent": [Circular],
                  "readonly": true,
                  "subTypes": Array [],
                },
                ColumnDataset {
                  "collection": [Circular],
                  "column": Object {
                    "name": "createdAt",
                    "type": "date",
                  },
                  "columns": Array [],
                  "index": undefined,
                  "parent": [Circular],
                  "readonly": true,
                  "subTypes": Array [],
                },
                ColumnDataset {
                  "collection": [Circular],
                  "column": Object {
                    "name": "updatedAt",
                    "type": "date",
                  },
                  "columns": Array [],
                  "index": undefined,
                  "parent": [Circular],
                  "readonly": true,
                  "subTypes": Array [],
                },
              ],
              "idType": "objectId",
              "indexes": Array [],
              "name": "collection2",
              "parent": [Circular],
              "specialColumns": Array [
                Array [
                  "createdAt",
                  "date",
                ],
                Array [
                  "updatedAt",
                  "date",
                ],
                Array [
                  "_id",
                  "objectId",
                ],
              ],
            },
            [Circular],
          ],
          "parent": [MockFunction],
          "path": "path",
        },
        "specialColumns": Array [
          Array [
            "createdAt",
            "date",
          ],
          Array [
            "updatedAt",
            "date",
          ],
          Array [
            "_id",
            "objectId",
          ],
        ],
      },
    },
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m──────────────[22m",
    "type": "separator",
  },
  Object {
    "name": "Create new collection",
    "short": "[31mCommand - Create collection[39m",
    "value": Object {
      "action": "create",
    },
  },
  Object {
    "name": "Remove group",
    "short": "[31mCommand - Remove group[39m",
    "value": Object {
      "action": "remove",
      "data": undefined,
    },
  },
  Object {
    "name": "Save",
    "short": "[31mCommand - Save[39m",
    "value": Object {
      "action": "save",
    },
  },
  Object {
    "name": "Save and Write",
    "short": "[31mCommand - Write[39m",
    "value": Object {
      "action": "write",
    },
  },
  Object {
    "name": "Back",
    "short": "[31mCommand - Back[39m",
    "value": Object {
      "action": "back",
    },
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
]
`;

exports[`Check the GroupMenu class it should be return menu selection when exec() is called without collections 1`] = `
Array [
  Separator {
    "line": "[2mGroup: [0m[1mpath[2m[0m[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
  Separator {
    "line": "[2mCollections list[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m- No collections defined -[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m──────────────[22m",
    "type": "separator",
  },
  Object {
    "name": "Create new collection",
    "short": "[31mCommand - Create collection[39m",
    "value": Object {
      "action": "create",
    },
  },
  Object {
    "name": "Remove group",
    "short": "[31mCommand - Remove group[39m",
    "value": Object {
      "action": "remove",
      "data": undefined,
    },
  },
  Object {
    "name": "Save",
    "short": "[31mCommand - Save[39m",
    "value": Object {
      "action": "save",
    },
  },
  Object {
    "name": "Save and Write",
    "short": "[31mCommand - Write[39m",
    "value": Object {
      "action": "write",
    },
  },
  Object {
    "name": "Back",
    "short": "[31mCommand - Back[39m",
    "value": Object {
      "action": "back",
    },
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
]
`;
